e2e-test/src/main.rs

```diff
@@ -0,0 +1,245 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{env::VarError, future::Future};
+
+use astarte_device_sdk::prelude::*;
+use eyre::{bail, ensure, eyre, Context, OptionExt};
+use interfaces::ServerAggregate;
+use tempfile::tempdir;
+use tokio::task::JoinSet;
+use tracing::{debug, error, instrument};
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt, EnvFilter};
+use uuid::{uuid, Uuid};
+
+use crate::{
+    api::Api,
+    device_sdk::{init_node, Node},
+    interfaces::{
+        DeviceAggregate, DeviceDatastream, DeviceProperty, ServerDatastream, ServerProperty,
+        ENDPOINTS, INTERFACE_NAMES,
+    },
+    message_hub::{init_message_hub, MsgHub},
+};
+
+pub mod api;
+pub mod device_sdk;
+pub mod interfaces;
+pub mod message_hub;
+pub mod utils;
+
+pub const GRPC_PORT: u16 = 50051;
+pub const UUID: Uuid = uuid!("acc78dae-194c-4942-8f33-9f719629e316");
+
+fn env_filter() -> eyre::Result<EnvFilter> {
+    let filter = std::env::var("RUST_LOG").or_else(|err| match err {
+        VarError::NotPresent => Ok(
+            "e2e_test=trace,astarte_message_hub=debug,astarte_device_sdk=debug,tower_http=debug"
+                .to_string(),
+        ),
+        err @ VarError::NotUnicode(_) => Err(err),
+    })?;
+
+    let env_filter = EnvFilter::try_new(filter)?;
+
+    Ok(env_filter)
+}
+
+#[tokio::main]
+async fn main() -> eyre::Result<()> {
+    color_eyre::install()?;
+
+    let filter = env_filter()?;
+    tracing_subscriber::registry()
+        .with(tracing_subscriber::fmt::layer())
+        .with(filter)
+        .try_init()?;
+
+    let dir = tempdir()?;
+
+    let api = Api::try_from_env()?;
+
+    let mut tasks = JoinSet::new();
+
+    let msghub = init_message_hub(dir.path(), &mut tasks).await?;
+    let node = init_node(&mut tasks).await?;
+
+    tasks.spawn(async move { e2e_test(api, msghub, node).await });
+
+    while let Some(res) = tasks.join_next().await {
+        match res {
+            Ok(res) => {
+                res.wrap_err("task failed")?;
+            }
+            Err(err) if err.is_cancelled() => {}
+            Err(err) => {
+                return Err(err).wrap_err("couldn't join task");
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Retry the future multiple times
+async fn retry<F, T, U>(times: usize, mut f: F) -> eyre::Result<U>
+where
+    F: FnMut() -> T,
+    T: Future<Output = eyre::Result<U>>,
+{
+    for i in 1..=times {
+        match (f)().await {
+            Ok(o) => return Ok(o),
+            Err(err) => {
+                error!("failed retry {i} for: {err}")
+            }
+        }
+    }
+
+    bail!("to many attempts")
+}
+
+#[instrument(skip_all)]
+async fn e2e_test(api: Api, msghub: MsgHub, mut node: Node) -> eyre::Result<()> {
+    // Check that the attach worked by checking the message hub interfaces
+    retry(4, || async {
```

Comment:
I would introduce a cont variable rather than embedding it into the function

Reply:
I don't think a constant will make the retry clearer. Since it used only here, it will just mask the value and make people go searching for it.

It's also just an empirical value to make the test more robust, so it has little value.